File: backend-go/internal/domain/glb_parity_canonical_test.go
package domain_test

import (
	"encoding/json"
	"math"
	"os"
	"testing"

	"github.com/tiagofur/muebles-backend/internal/domain"
)

// Canonical SKP/GLB parity (#669 Gate P0).
//
// Single numeric authority: contracts/fixtures/glb-parity-canonical.json.
// The Go domain must derive the same asset normalization and compose the same
// placement chain (furniture x assembly x member x inverse(mountFrame)) that
// the SketchUp host smoke measured and the TS domain suite asserts, so no
// runtime can keep a "slightly different" interpretation of the GLB boundary.
//
// The expected world points are hand-derived constants in the canonical file;
// they are never produced by the functions under test.

const glbParityFixturePath = "../../../contracts/fixtures/glb-parity-canonical.json"
const glbParityToleranceMm = 1e-3
const glbParityMutantMinErrorMm = 1.0

type glbParityBasis struct {
	X [3]float64 `json:"x"`
	Y [3]float64 `json:"y"`
	Z [3]float64 `json:"z"`
}

type glbParityFixture struct {
	Asset struct {
		NominalExtentsMm    [3]float64          `json:"nominalExtentsMm"`
		MountFrame          domain.HardwareMountFrame `json:"mountFrame"`
		ReferencePointsAssetMm [][3]float64      `json:"referencePointsAssetMm"`
	} `json:"asset"`
	GlbRepresentation struct {
		SourceUnits               string      `json:"sourceUnits"`
		UpAxis                    string      `json:"upAxis"`
		ExpectedGlbReferencePointsM [][3]float64 `json:"expectedGlbReferencePointsM"`
	} `json:"glbRepresentation"`
	PlacementChain struct {
		Furniture struct {
			TranslationMm [3]float64 `json:"translationMm"`
		} `json:"furniture"`
		Assembly struct {
			TranslationMm [3]float64 `json:"translationMm"`
		} `json:"assembly"`
		Member struct {
			TranslationMm [3]float64    `json:"translationMm"`
			Basis         glbParityBasis `json:"basis"`
		} `json:"member"`
	} `json:"placementChain"`
	Expected struct {
		AssetNormalization struct {
			TranslationMm [3]float64    `json:"translationMm"`
			Basis         glbParityBasis `json:"basis"`
		} `json:"assetNormalization"`
		ReferencePoints []struct {
			AssetMm        [3]float64 `json:"assetMm"`
			ExpectedWorldMm [3]float64 `json:"expectedWorldMm"`
		} `json:"referencePoints"`
	} `json:"expected"`
}

func loadGlbParityFixture(t *testing.T) glbParityFixture {
	t.Helper()
	raw, err := os.ReadFile(glbParityFixturePath)
	if err != nil {
		t.Fatalf("read canonical fixture: %v", err)
	}
	var fixture glbParityFixture
	if err := json.Unmarshal(raw, &fixture); err != nil {
		t.Fatalf("parse canonical fixture: %v", err)
	}
	return fixture
}

func applyBasis(b glbParityBasis, p [3]float64) [3]float64 {
	return [3]float64{
		p[0]*b.X[0] + p[1]*b.Y[0] + p[2]*b.Z[0],
		p[0]*b.X[1] + p[1]*b.Y[1] + p[2]*b.Z[1],
		p[0]*b.X[2] + p[1]*b.Y[2] + p[2]*b.Z[2],
	}
}

func addPoints(a, b [3]float64) [3]float64 {
	return [3]float64{a[0] + b[0], a[1] + b[1], a[2] + b[2]}
}

func pointDistanceMm(a, b [3]float64) float64 {
	dx, dy, dz := a[0]-b[0], a[1]-b[1], a[2]-b[2]
	return math.Sqrt(dx*dx + dy*dy + dz*dz)
}

// glbPointToAssetMm mirrors the canonical GLB boundary conversion
// (metres Y-up -> mm Z-up): asset = (X, -Z, Y) * unitScale.
func glbPointToAssetMm(sourceUnits string, upAxis string, p [3]float64) [3]float64 {
	scale := 1000.0
	switch sourceUnits {
	case "mm":
		scale = 1.0
	case "cm":
		scale = 10.0
	case "inch":
		scale = 25.4
	}
	x, y, z := p[0]*scale, p[1]*scale, p[2]*scale
	if upAxis == "z" {
		return [3]float64{x, y, z}
	}
	return [3]float64{x, -z, y}
}

// chainToWorld composes furniture x assembly x (member x normalization).
func chainToWorld(f glbParityFixture, norm domain.HardwareAssetNormalization, p [3]float64) [3]float64 {
	normalized := norm.Apply(p)
	member := addPoints(f.PlacementChain.Member.TranslationMm, applyBasis(f.PlacementChain.Member.Basis, normalized))
	assembly := addPoints(f.PlacementChain.Assembly.TranslationMm, member)
	return addPoints(f.PlacementChain.Furniture.TranslationMm, assembly)
}

func TestGlbParityCanonicalNormalizationMatchesContract(t *testing.T) {
	fixture := loadGlbParityFixture(t)
	norm, err := domain.DeriveAssetNormalization(fixture.Asset.MountFrame)
	if err != nil {
		t.Fatalf("derive normalization: %v", err)
	}
	want := fixture.Expected.AssetNormalization
	for i := 0; i < 3; i++ {
		if math.Abs(norm.TranslationMm[i]-want.TranslationMm[i]) > 1e-9 {
			t.Fatalf("normalization translation[%d] = %v, want %v", i, norm.TranslationMm[i], want.TranslationMm[i])
		}
	}
	bases := map[string][2][3]float64{
		"x": {norm.Basis.X, want.Basis.X},
		"y": {norm.Basis.Y, want.Basis.Y},
		"z": {norm.Basis.Z, want.Basis.Z},
	}
	for axis, pair := range bases {
		for i := 0; i < 3; i++ {
			if math.Abs(pair[0][i]-pair[1][i]) > 1e-9 {
				t.Fatalf("normalization basis %s[%d] = %v, want %v", axis, i, pair[0][i], pair[1][i])
			}
		}
	}
}

func TestGlbParityCanonicalWorldPoints(t *testing.T) {
	fixture := loadGlbParityFixture(t)
	norm, err := domain.DeriveAssetNormalization(fixture.Asset.MountFrame)
	if err != nil {
		t.Fatalf("derive normalization: %v", err)
	}

	if len(fixture.GlbRepresentation.ExpectedGlbReferencePointsM) != len(fixture.Expected.ReferencePoints) {
		t.Fatalf("canonical fixture glb/reference point count mismatch")
	}

	for i, rp := range fixture.Expected.ReferencePoints {
		assetFromGlb := glbPointToAssetMm(
			fixture.GlbRepresentation.SourceUnits,
			fixture.GlbRepresentation.UpAxis,
			fixture.GlbRepresentation.ExpectedGlbReferencePointsM[i],
		)
		for axis := 0; axis < 3; axis++ {
			if math.Abs(assetFromGlb[axis]-rp.AssetMm[axis]) > 1e-6 {
				t.Fatalf("P%d glb->asset axis %d = %v, want %v", i, axis, assetFromGlb[axis], rp.AssetMm[axis])
			}
		}
		world := chainToWorld(fixture, norm, assetFromGlb)
		if delta := pointDistanceMm(world, rp.ExpectedWorldMm); delta > glbParityToleranceMm {
			t.Fatalf("P%d world = %v, want %v (delta %g mm)", i, world, rp.ExpectedWorldMm, delta)
		}
	}
}

func TestGlbParityCanonicalMutantSensitivity(t *testing.T) {
	fixture := loadGlbParityFixture(t)
	norm, err := domain.DeriveAssetNormalization(fixture.Asset.MountFrame)
	if err != nil {
		t.Fatalf("derive normalization: %v", err)
	}

	memberNoNorm := func(p [3]float64) [3]float64 {
		member := addPoints(fixture.PlacementChain.Member.TranslationMm, applyBasis(fixture.PlacementChain.Member.Basis, p))
		return addPoints(addPoints(fixture.PlacementChain.Furniture.TranslationMm, fixture.PlacementChain.Assembly.TranslationMm), member)
	}
	mutants := map[string]func(i int) [3]float64{
		"unit_x25_4": func(i int) [3]float64 {
			p := fixture.GlbRepresentation.ExpectedGlbReferencePointsM[i]
			scaled := [3]float64{p[0] * 25.4, p[1] * 25.4, p[2] * 25.4}
			return chainToWorld(fixture, norm, glbPointToAssetMm(fixture.GlbRepresentation.SourceUnits, fixture.GlbRepresentation.UpAxis, scaled))
		},
		"unit_x1000": func(i int) [3]float64 {
			p := fixture.GlbRepresentation.ExpectedGlbReferencePointsM[i]
			scaled := [3]float64{p[0] * 1000, p[1] * 1000, p[2] * 1000}
			return chainToWorld(fixture, norm, glbPointToAssetMm(fixture.GlbRepresentation.SourceUnits, fixture.GlbRepresentation.UpAxis, scaled))
		},
		"omitted": func(i int) [3]float64 {
			asset := fixture.Asset.ReferencePointsAssetMm[i]
			return memberNoNorm(asset)
		},
		"double": func(i int) [3]float64 {
			asset := fixture.Asset.ReferencePointsAssetMm[i]
			return chainToWorld(fixture, norm, norm.Apply(asset))
		},
	}

	for mode, mutant := range mutants {
		minDelta := math.Inf(1)
		for i, rp := range fixture.Expected.ReferencePoints {
			if delta := pointDistanceMm(mutant(i), rp.ExpectedWorldMm); delta < minDelta {
				minDelta = delta
			}
		}
		if minDelta <= glbParityMutantMinErrorMm {
			t.Fatalf("mutant %s would not be detectable (min error %g mm)", mode, minDelta)
		}
	}
}

File: packages/domain/src/glbParityCanonical.test.ts
import { readFileSync } from 'node:fs';
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';
import { describe, expect, it } from 'vitest';

import {
  type AssemblyBasis,
  type AssemblyMemberTransform,
  type HardwareMountFrame,
  composeMemberTransform,
  deriveAssetNormalization,
} from './agregadoAssembly';
import {
  type GlbCoordinateSpace,
  applyAssemblyPlacement,
  collectGlbVertices,
  glbPointToAssetMm,
  multiplyAssemblyPlacements,
  nearestGlbVertex,
  parseGlbContainer,
  validateGlbSelfContainedPolicy,
} from './glbRepresentation';

// --- #669 Gate P0: SKP <-> GLB parity (canonical contract) ------------------
//
// Single numeric authority: contracts/fixtures/glb-parity-canonical.json.
// The committed GLB (contracts/fixtures/glb-parity-bracket.glb) carries the
// bytes produced by the SketchUp host smoke through GlbWriter; this suite
// reads those exact bytes, maps them back to asset mm through the declared
// unit/axis conversion (the ONLY GLB-boundary normalization) and pushes them
// through the production chain
//
//   world = T_furniture x T_assembly x T_member x inverse(T_mountFrame) x P
//
// The expected world points are hand-derived constants in the canonical file
// (independent arithmetic, also equal to the #668 host-measured evidence), so
// this suite fails loudly under any drift between SketchUp and the web path.

const root = join(dirname(fileURLToPath(import.meta.url)), '../../..');
const canonicalPath = join(root, 'contracts/fixtures/glb-parity-canonical.json');
const glbPath = join(root, 'contracts/fixtures/glb-parity-bracket.glb');

interface CanonicalFixture {
  readonly asset: {
    readonly nominalExtentsMm: readonly [number, number, number];
    readonly referencePointsAssetMm: readonly (readonly [number, number, number])[];
    readonly mountFrame: HardwareMountFrame;
  };
  readonly glbRepresentation: {
    readonly sourceUnits: 'mm' | 'cm' | 'm' | 'inch';
    readonly upAxis: 'y' | 'z';
    readonly expectedGlbReferencePointsM: readonly (readonly [number, number, number])[];
  };
  readonly placementChain: {
    readonly furniture: { readonly translationMm: readonly [number, number, number] };
    readonly assembly: { readonly translationMm: readonly [number, number, number] };
    readonly member: AssemblyMemberTransform;
  };
  readonly expected: {
    readonly assetNormalization: {
      readonly translationMm: readonly [number, number, number];
      readonly basis: AssemblyBasis;
    };
    readonly effectiveMemberTransform: AssemblyMemberTransform;
    readonly referencePoints: readonly {
      readonly assetMm: readonly [number, number, number];
      readonly expectedWorldMm: readonly [number, number, number];
    }[];
    readonly pairwiseDistancesMm: Readonly<Record<string, number>>;
    readonly rigidity: { readonly scale: readonly number[]; readonly determinant: number };
    readonly tolerance: {
      readonly worldPointToleranceMm: number;
      readonly glbVertexMatchToleranceM: number;
      readonly mutantMinErrorMm: number;
    };
  };
}

const canonical = JSON.parse(readFileSync(canonicalPath, 'utf8')) as CanonicalFixture;
const glbBytes = new Uint8Array(readFileSync(glbPath));
const glbSpace: GlbCoordinateSpace = {
  sourceUnits: canonical.glbRepresentation.sourceUnits,
  upAxis: canonical.glbRepresentation.upAxis,
};
const identityBasis: AssemblyBasis = {
  x: [1, 0, 0],
  y: [0, 1, 0],
  z: [0, 0, 1],
};
const tolerance = canonical.expected.tolerance;

function worldChain(): ReturnType<typeof multiplyAssemblyPlacements> {
  const effective = composeMemberTransform(
    canonical.placementChain.member,
    canonical.asset.mountFrame,
  );
  const assemblyPlacement = multiplyAssemblyPlacements(
    {
      translationMm: canonical.placementChain.assembly.translationMm,
      basis: identityBasis,
    },
    effective,
  );
  return multiplyAssemblyPlacements(
    {
      translationMm: canonical.placementChain.furniture.translationMm,
      basis: identityBasis,
    },
    assemblyPlacement,
  );
}

function glbVertexWorldPoints(): readonly (readonly [number, number, number])[] {
  const parsed = parseGlbContainer(glbBytes);
  const issues = validateGlbSelfContainedPolicy(parsed);
  expect(issues).toEqual([]);
  const vertices = collectGlbVertices(parsed);
  return canonical.glbRepresentation.expectedGlbReferencePointsM.map((expectedGlb) => {
    const vertex = nearestGlbVertex(vertices, expectedGlb);
    const distance = Math.hypot(
      vertex[0] - expectedGlb[0],
      vertex[1] - expectedGlb[1],
      vertex[2] - expectedGlb[2],
    );
    expect(distance).toBeLessThan(tolerance.glbVertexMatchToleranceM);
    return glbPointToAssetMm(glbSpace, vertex);
  });
}

describe('#669 canonical SKP/GLB parity', () => {
  it('committed GLB is a self-contained container matching the canonical contract', () => {
    const parsed = parseGlbContainer(glbBytes);
    expect(validateGlbSelfContainedPolicy(parsed)).toEqual([]);
    expect(parsed.json.asset?.version).toBe('2.0');
    expect(parsed.json.buffers?.[0]?.uri ?? '').toBe('');
    expect((parsed.json.meshes ?? []).length).toBeGreaterThan(0);
  });

  it('derived asset normalization equals the hand-derived contract constants', () => {
    const norm = deriveAssetNormalization(canonical.asset.mountFrame);
    expect([...norm.translationMm]).toEqual([...canonical.expected.assetNormalization.translationMm]);
    expect([...norm.basis.x]).toEqual([...canonical.expected.assetNormalization.basis.x]);
    expect([...norm.basis.y]).toEqual([...canonical.expected.assetNormalization.basis.y]);
    expect([...norm.basis.z]).toEqual([...canonical.expected.assetNormalization.basis.z]);
  });

  it('effective member transform (T_member x inverse(T_mountFrame)) matches the contract', () => {
    const effective = composeMemberTransform(
      canonical.placementChain.member,
      canonical.asset.mountFrame,
    );
    const expectedEffective = canonical.expected.effectiveMemberTransform;
    expect([...effective.translationMm]).toEqual([...expectedEffective.translationMm]);
    expect([...effective.basis.x]).toEqual([...expectedEffective.basis.x]);
    expect([...effective.basis.y]).toEqual([...expectedEffective.basis.y]);
    expect([...effective.basis.z]).toEqual([...expectedEffective.basis.z]);
  });

  it('reference points read from the committed GLB land on the hand-derived world constants', () => {
    const world = worldChain();
    const assetPoints = glbVertexWorldPoints();
    expect(assetPoints.length).toBe(canonical.expected.referencePoints.length);

    canonical.expected.referencePoints.forEach((expected, index) => {
      const worldPoint = applyAssemblyPlacement(world, assetPoints[index]!);
      const delta = Math.hypot(
        worldPoint[0] - expected.expectedWorldMm[0],
        worldPoint[1] - expected.expectedWorldMm[1],
        worldPoint[2] - expected.expectedWorldMm[2],
      );
      expect(delta).toBeLessThan(tolerance.worldPointToleranceMm);
    });
  });

  it('dimensions and pairwise distances are preserved through the whole chain', () => {
    const parsed = parseGlbContainer(glbBytes);
    const vertices = collectGlbVertices(parsed);
    const assetVertices = [] as [number, number, number][];
    for (let i = 0; i + 2 < vertices.length; i += 3) {
      assetVertices.push([
        vertices[i],
        vertices[i + 1],
        vertices[i + 2],
      ] as [number, number, number]);
    }
    const extents = ([0, 1, 2] as const).map((axis) => {
      const values = assetVertices.map((v) => glbPointToAssetMm(glbSpace, v)[axis]);
      return Math.max(...values) - Math.min(...values);
    });
    canonical.asset.nominalExtentsMm.forEach((nominal, axis) => {
      expect(Math.abs(extents[axis]! - nominal)).toBeLessThan(tolerance.worldPointToleranceMm);
    });

    const world = worldChain();
    const assetPoints = glbVertexWorldPoints();
    const worldPoints = assetPoints.map((p) => applyAssemblyPlacement(world, p));
    const distance = (a: readonly number[], b: readonly number[]) =>
      Math.hypot(a[0]! - b[0]!, a[1]! - b[1]!, a[2]! - b[2]!);
    const distances = canonical.expected.pairwiseDistancesMm;
    const [d01, d12, d02] = [distances['p0p1'], distances['p1p2'], distances['p0p2']];
    expect(d01, 'canonical pairwise p0p1').toBeDefined();
    expect(d12, 'canonical pairwise p1p2').toBeDefined();
    expect(d02, 'canonical pairwise p0p2').toBeDefined();
    expect(distance(worldPoints[0]!, worldPoints[1]!)).toBeCloseTo(d01!, 3);
    expect(distance(worldPoints[1]!, worldPoints[2]!)).toBeCloseTo(d12!, 3);
    expect(distance(worldPoints[0]!, worldPoints[2]!)).toBeCloseTo(d02!, 3);
  });

  it('world chain stays rigid: scale [1,1,1], determinant +1, no shear, no mirror', () => {
    const world = worldChain();
    const det =
      world.basis.x[0] * (world.basis.y[1] * world.basis.z[2] - world.basis.y[2] * world.basis.z[1]) -
      world.basis.x[1]! * (world.basis.y[0]! * world.basis.z[2]! - world.basis.y[2]! * world.basis.z[0]!) +
      world.basis.x[2]! * (world.basis.y[0]! * world.basis.z[1]! - world.basis.y[1]! * world.basis.z[0]!);
    expect(det).toBeCloseTo(canonical.expected.rigidity.determinant, 6);
    const dot = (a: readonly number[], b: readonly number[]) =>
      a[0]! * b[0]! + a[1]! * b[1]! + a[2]! * b[2]!;
    expect(dot(world.basis.x, world.basis.y)).toBeCloseTo(0, 6);
    expect(dot(world.basis.x, world.basis.z)).toBeCloseTo(0, 6);
    expect(dot(world.basis.y, world.basis.z)).toBeCloseTo(0, 6);
    expect(Math.hypot(...world.basis.x)).toBeCloseTo(1, 6);
    expect(Math.hypot(...world.basis.y)).toBeCloseTo(1, 6);
    expect(Math.hypot(...world.basis.z)).toBeCloseTo(1, 6);
  });

  it('unit and normalization defects are detectable (mutant sensitivity)', () => {
    const expectedWorlds = canonical.expected.referencePoints.map((rp) => rp.expectedWorldMm);
    const assetPoints = glbVertexWorldPoints();
    const member = canonical.placementChain.member;
    const mountFrame = canonical.asset.mountFrame;

    const wrapWithFurnitureAssembly = (memberTransform: AssemblyMemberTransform) =>
      multiplyAssemblyPlacements(
        {
          translationMm: canonical.placementChain.furniture.translationMm,
          basis: identityBasis,
        },
        multiplyAssemblyPlacements(
          {
            translationMm: canonical.placementChain.assembly.translationMm,
            basis: identityBasis,
          },
          memberTransform,
        ),
      );

    const wrongUnitAssetPoints = (scale: number) =>
      canonical.glbRepresentation.expectedGlbReferencePointsM.map(
        (p) => glbPointToAssetMm(glbSpace, [p[0]! * scale, p[1]! * scale, p[2]! * scale]) as [number, number, number],
      );

    // Chain variants: each mutant mutates exactly one link of the composition.
    const normalizationWithBasis = (point: readonly number[], basis: AssemblyBasis) => {
      const origin = mountFrame.originMm;
      const delta = [point[0]! - origin[0], point[1]! - origin[1], point[2]! - origin[2]];
      const dotAxes = (a: readonly number[], b: readonly number[]) =>
        a[0]! * b[0]! + a[1]! * b[1]! + a[2]! * b[2]!;
      return [dotAxes(basis.x, delta), dotAxes(basis.y, delta), dotAxes(basis.z, delta)];
    };
    const transposedMountBasis: AssemblyBasis = {
      x: [mountFrame.basis.x[0], mountFrame.basis.y[0], mountFrame.basis.z[0]],
      y: [mountFrame.basis.x[1], mountFrame.basis.y[1], mountFrame.basis.z[1]],
      z: [mountFrame.basis.x[2], mountFrame.basis.y[2], mountFrame.basis.z[2]],
    };

    const mutants: Readonly<Record<string, { chain: ReturnType<typeof worldChain>; points: readonly (readonly [number, number, number])[] }>> = {
      unit_x25_4: { chain: worldChain(), points: wrongUnitAssetPoints(25.4) },
      unit_div25_4: { chain: worldChain(), points: wrongUnitAssetPoints(1 / 25.4) },
      unit_x1000: { chain: worldChain(), points: wrongUnitAssetPoints(1000) },
      unit_div1000: { chain: worldChain(), points: wrongUnitAssetPoints(0.001) },
      omitted: { chain: wrapWithFurnitureAssembly(member), points: assetPoints },
      double: {
        chain: wrapWithFurnitureAssembly(composeMemberTransform(composeMemberTransform(member, mountFrame), mountFrame)),
        points: assetPoints,
      },
      transposed: {
        chain: wrapWithFurnitureAssembly(member),
        points: assetPoints.map(
          (p) => normalizationWithBasis(p, transposedMountBasis) as [number, number, number],
        ),
      },
    };

    for (const [mode, { chain, points }] of Object.entries(mutants)) {
      const minError = Math.min(
        ...points.map((p, index) => {
          const expected = expectedWorlds[index]!;
          const worldPoint = applyAssemblyPlacement(chain, p);
          return Math.hypot(
            worldPoint[0]! - expected[0],
            worldPoint[1]! - expected[1],
            worldPoint[2]! - expected[2],
          );
        }),
      );
      expect(minError, `mutant ${mode} must be detectable`).toBeGreaterThan(
        tolerance.mutantMinErrorMm,
      );
    }
  });
});
